src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt

```diff
@@ -91,33 +91,49 @@ fun processItemDeclarations2(
  * We need to get [ModData] to check if we can use new resolve, which is not fast,
  * so we unite check and actual resolve as an optimization.
  */
-fun processMacros(scope: RsMod, processor: RsResolveProcessor, runBeforeResolve: () -> Boolean): Boolean? {
+fun processMacros(
+    scope: RsMod,
+    processor: RsResolveProcessor,
+    runBeforeResolve: (() -> Boolean)? = null
+): Boolean? {
     val (project, defMap, modData) = when (val info = getModInfo(scope)) {
         is CantUseNewResolve -> return null
         InfoNotFound -> return false
         is RsModInfo -> info
     }
-    if (runBeforeResolve()) return true
+    if (runBeforeResolve != null && runBeforeResolve()) return true
 
     return modData.processMacros(processor, defMap, project)
 }
 
 private fun ModData.processMacros(processor: RsResolveProcessor, defMap: CrateDefMap, project: Project): Boolean {
+    val processedMacros = hashSetOf<RsNamedElement>()
     for ((name, macroInfo) in legacyMacros.entriesWithName(processor.name)) {
         val visItem = VisItem(macroInfo.path, Visibility.Public)
-        val macros = visItem.toPsi(defMap, project, Namespace.Macros).singleOrNull() ?: continue
-        processor(name, macros) && return true
+        val macro = visItem.toPsi(defMap, project, Namespace.Macros).singleOrNull() ?: continue
+        processedMacros += macro
+        processor(name, macro) && return true
```

Comment:
Such constructions make me cry 😢

Reply:
I remembered to not use such constructions in new code, but this is existing code, that's why I didn't change it. Let's change it anyway

Reply:
Oh, indeed